src/mesh/parallel.jl

```diff
@@ -16,16 +19,38 @@ function partition!(mesh::ParallelTreeMesh)
   mesh.n_cells_by_rank = similar(n_leaves_per_rank)
 
   leaf_count = 0
-  last_id = leaves[n_leaves_per_rank[0]]
   mesh.first_cell_by_rank[0] = 1
-  mesh.n_cells_by_rank[0] = last_id
-  mesh.tree.mpi_ranks[1:last_id] .= 0
-  for d in 1:(length(n_leaves_per_rank)-1)
-    leaf_count += n_leaves_per_rank[d-1]
-    last_id = leaves[leaf_count + n_leaves_per_rank[d]]
-    mesh.first_cell_by_rank[d] = mesh.first_cell_by_rank[d-1] + mesh.n_cells_by_rank[d-1]
+  # Iterate over all ranks
+  for d in 0:(length(n_leaves_per_rank) - 1)
+    leaf_count += n_leaves_per_rank[d]
+    last_id = leaves[leaf_count]
+    parent_id = mesh.tree.parent_ids[last_id]
+
+    # Check if all children of the last parent are leaves
+    if all(id -> is_leaf(mesh.tree, id), mesh.tree.child_ids[:, parent_id]) && 
+        d < length(n_leaves_per_rank) - 1
+
+      # To keep children of parent together if they are all leaves, 
+      # all children are added to this rank
+      additional_cells = last_id+1:mesh.tree.child_ids[end, parent_id]
```

Comment:
That I don't understand: What if `mesh.tree.child_ids[end, parent_id]` is, e.g., `20`. Wouldn't that make `additional_cells = last_id + 1:20`?

Reply:
The `+` has a higher priority here, so it's `(last_id + 1):mesh.tree.child_ids[end, parent_id]`. So if `last_id` is a child of a parent whose children are all leaves, I want to add all these children and `additional_cells` gives me all ids greater than `last_id` that I want to add.